source/jormungandr/jormungandr/scenarios/new_default.py

```diff
@@ -810,30 +811,64 @@ def fill_journeys(self, request_type, api_request, instance):
-            _tag_by_mode(tmp_resp)
-            _tag_direct_path(tmp_resp)
-            _tag_bike_in_pt(tmp_resp)
-            journey_filter._filter_too_long_journeys(tmp_resp, request)
-            responses.extend(tmp_resp)  # we keep the error for building the response
-            if nb_journeys(tmp_resp) == 0:
+            new_resp = self.call_kraken(request_type, request, instance, krakens_call)
+            _tag_by_mode(new_resp)
+            _tag_direct_path(new_resp)
+            _tag_bike_in_pt(new_resp)
+            journey_filter._filter_too_late_journeys(new_resp, request)
+            if nb_journeys(new_resp) == 0:
                 # no new journeys found, we stop
+                # we still append the new_resp because there are journeys that a tagged as dead probably
+                responses.extend(new_resp)
                 break
 
-            request = self.create_next_kraken_request(request, tmp_resp)
-
-            # we filter unwanted journeys by side effects
-            journey_filter.filter_journeys(responses, instance, api_request)
-
-            #We allow one more call to kraken if there is no valid journey.
-            if nb_journeys(responses) == 0:
+            request = self.create_next_kraken_request(request, new_resp)
+
+            # we filter unwanted journeys in the new response
+            # note that filter_journeys returns a generator which will be evaluated later
+            filtered_new_resp = journey_filter.filter_journeys(new_resp, instance, api_request)
+
+            # duplicate the generator
+            tmp1, tmp2 = itertools.tee(filtered_new_resp)
+            qualified_journeys = journey_filter.get_qualified_journeys(responses)
+
+            # now we want to filter similar journeys in the new response which is done in 2 steps
+            # In the first step, we compare journeys from the new response only , 2 by 2
+            # hopefully, it may lead to some early return for the second step to improve the perf a little
+            # In the second step, we compare the journeys from the new response with those that have been qualified
+            # already in the former iterations
+            # note that the journeys_pool is a list of 2-element tuple of journeys
+            journeys_pool = itertools.chain(
+                # First step: compare journeys from the new response only
+                itertools.combinations(tmp1, 2),
+                # Second step:
+                # we use the itertools.product to create combinations between qualified journeys and new journeys
+                # Ex:
+                # new_journeys = [n_1, n_2]
+                # qualified_journeys = [q_1, q_2, q_3]
+                # itertools.product(new_journeys, qualified_journeys) gives combinations as follows:
+                # (n_1, q_1), (n_1, q_2),(n_1, q_3),(n_2, q_1),(n_2, q_2),(n_2, q_3)
+                itertools.product(tmp2, qualified_journeys),
```

Comment:
:confused: 
I'm not sure what's really going one here with this business logic, but when I see way more comments than actual code, my guts are telling me that something is fishy here.

Reply:
@nonifier  :D most of them are parts of the example